bin/steps/python

```diff
@@ -131,34 +131,48 @@ if [ ! "$SKIP_INSTALL" ]; then
   # Record for future reference.
   echo "$PYTHON_VERSION" > .heroku/python-version
   echo "$STACK" > .heroku/python-stack
-  FRESH_PYTHON=true
 
   hash -r
 fi
 
-# Heroku uses the get-pip utility maintained by the Python community to vendor Pip.
-# https://github.com/pypa/get-pip
-GETPIP_URL="https://lang-python.s3.amazonaws.com/etc/get-pip.py"
-GETPIP_PY="${TMPDIR:-/tmp}/get-pip.py"
+set -e
 
-if ! curl -s "${GETPIP_URL}" -o "$GETPIP_PY" &> /dev/null; then
-    mcount "failure.python.get-pip"
-    echo "Failed to pull down get-pip"
-    exit 1
+PIP_VERSION='20.1.1'
+# Must use setuptools <47.2.0 until we fix: https://github.com/heroku/heroku-buildpack-python/issues/1006
+SETUPTOOLS_VERSION='47.1.1'
+WHEEL_VERSION='0.34.2'
+
+if [[ "${PYTHON_VERSION}" == ${PY34}* ]]; then
+  # Python 3.4 support was dropped in pip 19.2, setuptools 44.0.0 and wheel 0.34.0.
+  PIP_VERSION='19.1.1'
+  SETUPTOOLS_VERSION='43.0.0'
+  WHEEL_VERSION='0.33.6'
+elif [[ "${PYTHON_VERSION}" == ${PY27}* || "${PYTHON_VERSION}" == ${PYPY27}* ]]; then
+  # Python 2.7 support was dropped in setuptools 45.0.0.
+  SETUPTOOLS_VERSION='44.1.1'
 fi
 
-# If a new Python has been installed or Pip isn't up to date:
-if [ "$FRESH_PYTHON" ] || [[ ! $(pip --version) == *$PIP_UPDATE* ]]; then
+puts-step "Installing pip ${PIP_VERSION}, setuptools ${SETUPTOOLS_VERSION} and wheel ${WHEEL_VERSION}"
 
-  puts-step "Installing pip"
+# We don't use get-pip.py, since:
+#  - it uses `--force-reinstall`, which is unnecessary here and slows down repeat builds
+#  - it means downloading pip twice (once embedded in get-pip.py, and again during
+#    the install, since get-pip.py can't install the embedded version directly)
+#  - we would still have to manage several versions of get-pip.py, to support older Pythons.
+# Instead, we use the pip wheel to install itself, using the method described here:
+# https://github.com/pypa/pip/issues/2351#issuecomment-69994524
 
-  # Remove old installations.
-  rm -fr /app/.heroku/python/lib/python*/site-packages/pip-*
-  rm -fr /app/.heroku/python/lib/python*/site-packages/setuptools-*
+PIP_WHEEL_FILENAME="pip-${PIP_VERSION}-py2.py3-none-any.whl"
+PIP_WHEEL_URL="https://lang-python.s3.amazonaws.com/common/${PIP_WHEEL_FILENAME}"
```

Comment:
Should this be using `$VENDOR_URL` rather than hardcoding the entire thing?

Reply:
`$VENDOR_URL` includes the `$STACK`, whereas these are not stack specific.

Refactoring the URL handling to separate out S3 base URL is on my list, but given the existing get-pip.py URL this PR replaces also hardcodes the URL, this seemed like something to do separately.

Reply:
Some context I forgot to mention:
* Currently `VENDOR_URL` is leaked to subprocesses such as `bin/post_compile`, so changing it has the chance of being a breaking change
* Adding an `S3_BASE_URL` and then using `VENDOR_URL="${S3_BASE_URL}/${STACK}` etc means that this feature would also really need rewriting/removing/...: https://github.com/heroku/heroku-buildpack-python/blob/d9b1c73f637f136eac038f526d161d83b94168d8/bin/compile#L43-L48

At the moment there are so many things to fix, many of which interact with each other, that I'm trying to avoid the temptation to scope creep each change too far.